server/lib/notifications/notification-model.ts

```diff
@@ -0,0 +1,146 @@
+import sql from 'sql-template-strings'
+import { NotificationType } from '../../../common/notifications'
+import db from '../db/index'
+import { Dbify } from '../db/types'
+
+export interface BaseNotificationData {
+  readonly type: NotificationType
+}
+
+export type NotificationData = BaseNotificationData
+
+export interface Notification {
+  id: string
+  userId: number
+  read: boolean
+  visible: boolean
+  createdAt: Date
+  data: NotificationData
+}
+
+type DbNotification = Dbify<Notification>
+
+function fromDbNotification(result: Readonly<DbNotification>): Notification {
+  return {
+    id: result.id,
+    userId: result.user_id,
+    read: result.read,
+    visible: result.visible,
+    createdAt: result.created_at,
+    data: result.data,
+  }
+}
+
+/**
+ * Retrieves the list of 100 visible notifications for a particular user by default, ordered by the
+ * time they were created.
+ */
+export async function retrieveNotifications({
+  userId,
+  type,
+  visible = true,
+  limit = 100,
+}: {
+  userId: number
+  type?: string
+  visible?: boolean
+  limit?: number
+}): Promise<Notification[]> {
+  const { client, done } = await db()
+  try {
+    const query = sql`
+      SELECT id, user_id, read, visible, created_at, data
+      FROM notifications
+      WHERE user_id = ${userId} AND visible = ${visible}
+    `
+
+    if (type !== undefined) {
+      query.append(sql` AND data->>'type' = ${type}`)
+    }
+
+    query.append(sql`
+      ORDER BY created_at DESC
+      LIMIT ${limit};
+    `)
+
+    const result = await client.query<DbNotification>(query)
+    return result.rows.map(n => fromDbNotification(n))
+  } finally {
+    done()
+  }
+}
+
+/**
+ * Creates a new notification with the given data and persists it to the DB. Notifications are
+ * created as unread, but visible by default.
+ */
+export async function addNotification({
+  userId,
+  data,
+  createdAt = new Date(),
+}: {
+  userId: number
+  data: NotificationData
+  createdAt?: Date
```

Comment:
Not totally clear what happens here if this is undefined? Does postgres use the default value for the column? Might be more sensible to just add a default value for this param of `Date.now()`

Reply:
And actually looking at the migration, this doesn't even have a default value there, so definitely should have one here :)

Reply:
The `createdAt` has a default value in the argument list, so it _should_ use that if an `undefined` value is sent. I'm not sure if I'm missing something?